src/System.Buffers.Experimental/System/Buffers/Text/BufferWriter.cs

```diff
@@ -0,0 +1,67 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System.Diagnostics;
+using System.Runtime.CompilerServices;
+
+namespace System.Buffers.Text
+{
+    public ref partial struct BufferWriter
+    {
+        Span<byte> _buffer;
+        int _written;
+
+        public ReadOnlySpan<byte> NewLine;
+        public Func<int, Memory<byte>> Enlarge;
+
+        static byte[] s_defaultNewline = new byte[] { (byte)'\n' };
+
+        public static BufferWriter Create(Span<byte> buffer) => new BufferWriter(buffer);
+        public static BufferWriter<TOutput> Create<TOutput>(TOutput output)
+            where TOutput : IOutput 
+            => new BufferWriter<TOutput>(output);
+
+        private BufferWriter(Span<byte> buffer)
+        {
+            _buffer = buffer;
+            _written = 0;
+            NewLine = s_defaultNewline;
+            Enlarge = null;
+        }
+
+        public Span<byte> Free => _buffer.Slice(_written);
+
+        public Span<byte> Written => _buffer.Slice(0, _written);
+
+        public int WrittenCount
+        {
+            get { return _written; }
+            set {
+                while (value > _buffer.Length) Resize(value);
+                _written = value;
+            }
+        }
+
+        public void Clear()
+            => _written = 0;
+        
+        private void Resize(int desiredBufferSize = 0)
+        {
+            if (Enlarge == null) throw new BufferTooSmallException();
+            if (desiredBufferSize <= 0) desiredBufferSize = _buffer.Length * 2 + 1;
```

Comment:
Do we add 1 here to make sure that if Length == 0, that we still increase the desiredBufferSize? If so, can you please add a comment.

Reply:
yes, and I will add the comment.